### ISSUE TITLE
Get this thrown at compile with ./configure --enable-hardening

### ISSUE THREAD

**Issue by Frankenmint** [bitcoin/bitcoin#7394]
At global scope:
cc1plus: warning: unrecognized command line option "-Wno-self-assign" [enabled by default]

**Comment by laanwj** [comment]
That's a clang-only warning, which is disabled, but doesn't exist on gcc. Perfectly harmless.

**Comment by 21E14** [comment]
'Tis confusing.

**Comment by paveljanik** [comment]
This is even an error on some older gccs as said in https://github.com/bitcoin/bitcoin/issues/7902#issuecomment-211515839

We have it hardcoded in `configure.ac`:
```
## TODO: Remove these hard-coded paths and flags. They are here for the sake of
##       compatibility with the legacy buildsystem.
##
if test "x$CXXFLAGS_overridden" = "xno"; then
  CXXFLAGS="$CXXFLAGS -Wall -Wextra -Wformat -Wformat-security -Wno-unused-parameter -Wno-self-assign"
fi
```

**Comment by theuni** [comment]
I wrote a fix for this a while back (as well as some that @morcos pointed out on osx), but I guess I never PR'd it. I'll see if I can dig that out.

**Comment by paveljanik** [comment]
@theuni OK. I have just started checking what warnings get printed by each of this -W.. So I'll wait :-)

**Comment by laanwj** [comment]
We really need this warning on clang though, I remember I added it because every single file was generating a warning (code generated by macros tends to have self-assignments).

**Comment by paveljanik** [comment]
FWIW: Removing `-Wno-self-assign` doesn't add new warnings here with clang 7.0.2 with the current master. I have just double-checked...

**Comment by laanwj** [comment]
OK, no clue then, in any case we need a mechanism for clang or gcc specific warnings.

**Comment by theuni** [comment]
@laanwj Yea, the trick is to run an autoconf check with "-Werror -Wno-self-assign" or so (we have macros for that). The annoying part is that gcc only warns on -Wno-foo when there's another warning present, so there a few other things needed. I had it nailed down to be a completely portable test, I believe. I'll dig it up.

**Comment by theuni** [comment]
Found this in my reflog: http://pastebin.com/raw/mGRe2umm . I'm unsure if that was the best version, will give it a try tomorrow. @paveljanik feel free to play with that.

**Comment by paveljanik** [comment]
`-Wno-unused-local-typedef` and `-Wno-deprecated-register` are new compared to the current master. Do we need them at all?

I commented them out to make comparison of build logs easier and the only change here is `configure` output:
```
checking whether C++ compiler accepts -Werror... yes
checking whether C++ compiler accepts -Wall... yes
checking whether C++ compiler accepts -Wextra... yes
checking whether C++ compiler accepts -Wformat... yes
checking whether C++ compiler accepts -Wformat-security... yes
checking whether C++ compiler accepts -Wunused-parameter... yes
checking whether C++ compiler accepts -Wself-assign... yes
```

This should fix this issue and also part of #7902. Thanks!

**Comment by theuni** [comment]
The deprecated-register was necessary when testing c++11, we can add it later if it comes up again. I believe the unused-local-typedef comes from boost in some cases. That might've been the one bothering @morcos.

**Comment by paveljanik** [comment]
But even adding them is no problem here. Will you PR the change?

**Comment by theuni** [comment]
Yes, will do.

**Comment by laanwj** [comment]
Closing, fixed by #7954
